locale_gen: fix: use glibc_ubuntu mode only if files for that exists

k0ste opened the pull request:
##### SUMMARY
Fixes #9734

##### ISSUE TYPE
<!--- Pick one or more below and delete the rest.
      'Test Pull Request' is for PRs that add/extend tests without code changes. -->
- Bugfix Pull Request

##### COMPONENT NAME
<!--- Write the SHORT NAME of the module, plugin, task or feature below. -->
locale_gen

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
This PR fixes issue for another glibc distro. The implementation should remove ubuntu_mode at all, and use `localectl list-locales` for any systemd distro in present days IMO

felixfontein commented on changelogs/fragments/9735-locale_gen-use_glibc_ubuntu_mode_only_if_file_exists.yml:
```diff
@@ -0,0 +1,2 @@
+bugfixes:
+  - "locale_gen - use glibc_ubuntu mode only if files for that exists (https://github.com/ansible-collections/community.general/pull/9735)."
```
```suggestion
  - "locale_gen - use ``glibc_ubuntu`` mode only if files for that exists (https://github.com/ansible-collections/community.general/pull/9735)."
```

felixfontein commented on plugins/modules/locale_gen.py:
```diff
@@ -119,13 +120,20 @@ def __init_module__(self):
+        if os.path.exists(ETC_LOCALE_GEN) and os.path.exists(SUPPORTED_LOCALES):
+            self.vars.ubuntu_mode = False
+            self.vars.mechanism = "ubuntu_glibc"
+        elif os.path.exists(ETC_LOCALE_GEN) and not os.path.exists(SUPPORTED_LOCALES):
```
The `and` is no longer necessary, since the previous `if` implies this:
```suggestion
        elif os.path.exists(ETC_LOCALE_GEN):
```

felixfontein reviewed (commented):
Thanks for your contribution!

CC @russoz

felixfontein commented:
For reference, the PR that refactored the module: #9238; the issue in which the main discussions happened: #9131.

k0ste commented:
> I am the one who introduced those changes and, at the time, it was understood that the module supported only Ubuntu and/or Debian systems, so I am caught a bit surprised and unaware that it has been used in other systems as well. Keep in mind that the docs and the tests for the module support only Debian/Ubuntu and have been like that for a while long time - long before I revamped the module.
> 
> On a side note: I noticed you mentioned you have been using this for 8 years, but your system is CentOS Stream 9, so I assume that worked in Stream 8 and maybe even CentOS 7 before that?

The original `locale_gen`, as I remember, released with Ansible 1.6. There are documentation

```python
#!/usr/bin/python
# -*- coding: utf-8 -*-

import os
import os.path
from subprocess import Popen, PIPE, call

DOCUMENTATION = '''
---
module: locale_gen
short_description: Creates of removes locales.
description:
     - Manages locales by editing /etc/locale.gen and invoking locale-gen.
version_added: "1.6"
options:
    name:
        description:
             - Name and encoding of the locale, such as "en_GB.UTF-8".
        required: true
        default: null
        aliases: []
    state:
      description:
           - Whether the locale shall be present.
      required: false
      choices: ["present", "absent"]
      default: "present"
'''

EXAMPLES = '''
# Ensure a locale exists.
- locale_gen: name=de_CH.UTF-8 state=present
'''
```

The original version had no limitations, so we used the module in the c7, c8s, c9s and Archlinux

----

Adjusted `when` for tests. Expect tests to fail if the appropriate language packs are not installed

felixfontein commented on tests/integration/targets/locale_gen/tasks/main.yml:
```diff
@@ -10,7 +10,8 @@
   ansible.builtin.meta: end_play
-  when: ansible_distribution not in ('Ubuntu', 'Debian')
+  when:
+    - "ansible_distribution == 'MacOSX'"
```
macOS is already excluded (together with FreeBSD and AIX) by `aliases`, so you can also simply remove the task.

You need to fix the tests for all the new platforms the PR is running them on, though.

russoz commented:
> > I am the one who introduced those changes and, at the time, it was understood that the module supported only Ubuntu and/or Debian systems, so I am caught a bit surprised and unaware that it has been used in other systems as well. Keep in mind that the docs and the tests for the module support only Debian/Ubuntu and have been like that for a while long time - long before I revamped the module.
> > On a side note: I noticed you mentioned you have been using this for 8 years, but your system is CentOS Stream 9, so I assume that worked in Stream 8 and maybe even CentOS 7 before that?
> 
> The original `locale_gen`, as I remember, released with Ansible 1.6. There are documentation
> 
> ```python
> #!/usr/bin/python
> # -*- coding: utf-8 -*-
> 
> import os
> import os.path
> from subprocess import Popen, PIPE, call
> 
> DOCUMENTATION = '''
> ---
> module: locale_gen
> short_description: Creates of removes locales.
> description:
>      - Manages locales by editing /etc/locale.gen and invoking locale-gen.
> version_added: "1.6"
> options:
>     name:
>         description:
>              - Name and encoding of the locale, such as "en_GB.UTF-8".
>         required: true
>         default: null
>         aliases: []
>     state:
>       description:
>            - Whether the locale shall be present.
>       required: false
>       choices: ["present", "absent"]
>       default: "present"
> '''
> 
> EXAMPLES = '''
> # Ensure a locale exists.
> - locale_gen: name=de_CH.UTF-8 state=present
> '''
> ```
> 
> The original version had no limitations, so we used the module in the c7, c8s, c9s and Archlinux
> 
> Adjusted `when` for tests. Expect tests to fail if the appropriate language packs are not installed

Thanks for the git spelunking :-) I certainly did not check the original versions of the module, but I am pretty sure the latest docs did not mention any of that, and the tests were not being executed for any of those distros.

Thanks for adjusting the `when`, but as @felixfontein pointed out, the tests will need adjustments to be executed in those systems. Please do adjsut them accordingly.

k0ste commented:
**pt_BR ISO-8859-1** -> **pt_BR.ISO-8859-1**
```
[root@archlinux]# locale-gen
Generating locales...
  en_US.UTF-8... done
error: Bad entry 'pt_BR.ISO-8859-1 '
Generation complete.
[root@archlinux]# locale-gen
Generating locales...
  en_US.UTF-8... done
  pt_BR.ISO-8859-1... done
Generation complete.
```

k0ste commented:
> Thanks for the git spelunking :-) I certainly did not check the original versions of the module, but I am pretty sure the latest docs did not mention any of that, and the tests were not being executed for any of those distros.
> 
> Thanks for adjusting the `when`, but as @felixfontein pointed out, the tests will need adjustments to be executed in those systems. Please do adjsut them accordingly.

You are welcome. Tests added

felixfontein commented on changelogs/fragments/9735-locale_gen-use_glibc_ubuntu_mode_only_if_file_exists.yml:
```diff
@@ -0,0 +1,2 @@
+bugfixes:
+  - "locale_gen - use ``glibc`` mode only if files for that exists. In other case us ``other_glibc`` mode (https://github.com/ansible-collections/community.general/pull/9735)."
```
```suggestion
  - "locale_gen - use ``glibc`` mode only if files for that exists. In other case use ``other_glibc`` mode (https://github.com/ansible-collections/community.general/pull/9735)."
```

russoz commented on tests/integration/targets/locale_gen/tasks/non_debian.yml:
```diff
@@ -0,0 +1,52 @@
+      when:
+        - "ansible_os_family == 'Archlinux'"
+
+- name: "Prepare locale.gen file for RedHat or Archlinux"
+  ansible.builtin.template:
+    src: "locale_gen.j2"
+    dest: "/etc/locale.gen"
+    owner: "root"
+    group: "root"
+    mode: "0644"
+    force: "yes"
```
OK, this is the part where it becomes odd. 

The module (again, even before my rewrite) had code to check whether a locale definition was available or not. In modern Debian/Ubuntu, it does that by checking the contents of the `/usr/share/i18n/SUPPORTED` file. That file is managed by the OS when packages are (un)installed.

What this task here tells me (unless I'm getting it wrong) is that RH and Arch Linux systems **do not keep a list of available locale definitions**. If that is so, what this task does (and possibly what your playbooks have been doing for years) is nothing more than _cheating_ the method `assert_available()` into calculating that the locales we artificially write to that file (here using the template + var) are the **only ones available**.

If you come to think of it, it makes no sense that users were forced to _create a file to make the module think that something managed by the OS_ was actually there (when it was not).

Having said that, I am thinking that asserting available (or not) locale definitions should be controlled by a parameter that would allow us to skip that check in the code, say `skip_availability_check` (`False` by default for backward compatibility). I understand this is a whole new thing, so it should be implemented in its own separate PR. When that happens, this step could be removed and replaced with the param `skip_availability_check=True`. Code-wise there's probably more to it, I am not going deep into that idea right now - as I wrote, that's a future PR.

For the time being, though, this incurs in documentation updates:
* Not this exactly, but somewhat related, please rewrite the part of the docs that state the module only works for Debian/Ubuntu
* Then, after mentioning it works for RH and Arch as well, please make it clear to the users that they are responsible for populating the `/etc/locale.gen` with the locales they want to be available to the module. Make it clear that locales not found in that file will be refuted by the module, even if their corresponding language pack packages are installed.

k0ste replied:
> The module (again, even before my rewrite) had code to check whether a locale definition was available or not. In modern Debian/Ubuntu, it does that by checking the contents of the `/usr/share/i18n/SUPPORTED` file. That file is managed by the OS when packages are (un)installed.

The main value of the module for us is to make sure that the locales specified in the configuration actually exist. The module has helped us out more than once in this case if we forgot to install the corresponding packages.

> What this task here tells me (unless I'm getting it wrong) is that RH and Arch Linux systems **do not keep a list of available locale definitions**. If that is so, what this task does (and possibly what your playbooks have been doing for years) is nothing more than _cheating_ the method `assert_available()` into calculating that the locales we artificially write to that file (here using the template + var) are the **only ones available**.

In Archlinux you can:
* use [SUPPORTED](https://gitlab.archlinux.org/archlinux/packaging/packages/glibc/-/blob/main/PKGBUILD?ref_type=heads#L176) file
* install locales pregenerated (`pacman -S glibc-locales`)
* prepare locale.gen file and generate locale. See [documentation](https://wiki.archlinux.org/title/Locale)

In RedHat you can:
* install `glibc-all-langpacks`
* install `glibc-minimal-langpack`
* `glibc-langpack-<xx>` for single language support

In both distro, you can list usable locales via systemd

```console
[root@host]# localectl list-locales
C.UTF-8
ast_ES.UTF-8
en_AU.UTF-8
en_BW.UTF-8
en_CA.UTF-8
en_DK.UTF-8
en_GB.UTF-8
en_HK.UTF-8
en_IE.UTF-8
en_NZ.UTF-8
en_PH.UTF-8
en_SC.UTF-8
en_SG.UTF-8
en_US.UTF-8
en_ZA.UTF-8
en_ZW.UTF-8
pl_PL.UTF-8
```

----

> If that is so, what this task does (and possibly what your playbooks have been doing for years) is nothing more than _cheating_ the method `assert_available()` into calculating that the locales we artificially write to that file (here using the template + var) are the **only ones available**.

Definitely. Because of how the module was originally written, we simply adapted to how other distributions work. I described the meaning and value above.

felixfontein replied:
Arch Linux's installation instructions basically tell you to edit `/etc/locale.gen` and comment in the locales you want to have, and then run `locale-gen` (and then run `echo LANG=... > /etc/locale.conf` to select the default locale). (https://wiki.archlinux.org/title/Installation_guide#Localization)

Basically `/etc/locale.gen` already contains the list of all supported locales. There's no need to template this file, unless you want to make sure you have a specific state.

So basically I don't see why this task should run on Arch Linux.

russoz commented on tests/integration/targets/locale_gen/tasks/non_debian.yml:
```diff
@@ -0,0 +1,52 @@
+    - name: "Install locale packages for RedHat"
+      ansible.builtin.dnf:
+        name:
+          - "glibc-langpack-pl"
+          - "glibc-langpack-en"
+        state: "present"
+        update_cache: "yes"
+      when:
+        - "ansible_os_family == 'RedHat'"
+    - name: "Install locale packages for Archlinux"
+      community.general.pacman:
+        name: "glibc-locales"
```
The requirement to install these packages should be mentioned in the module documentation, for the benefit of new users.

k0ste replied:
This documented in 

* https://docs.redhat.com/en/documentation/red_hat_enterprise_linux/9/html/configuring_basic_system_settings/assembly_changing-basic-environment-settings_configuring-basic-system-settings#proc_configuring-the-system-locale_assembly_changing-basic-environment-settings

* https://wiki.archlinux.org/title/Locale


Feel free to add the inline commit, if this links definitely should be present in module

russoz reviewed (commented):
hi @k0ste, thanks for the updates. I have another round of comments on the module.

k0ste commented on tests/integration/targets/locale_gen/tasks/non_debian.yml:
```diff
@@ -0,0 +1,52 @@
+# GNU General Public License v3.0+ (see LICENSES/GPL-3.0-or-later.txt or https://www.gnu.org/licenses/gpl-3.0.txt)
+# SPDX-License-Identifier: GPL-3.0-or-later
+
+- name: "Non dnf RedHat systemd are not supported"
```
c7 is retired OS, the `yum` also too. Is surprise, that Ansible still have tests over c7

felixfontein replied:
As long as ansible-core versions supported by this collection still support CentOS 7, we're kind of supporting it.

k0ste commented:
Gentle reminder that this is a bugfix. I added all the suggested inline-changes to the code, tests, description of how administrators work with locales in different environments (nobody likes it), I also added to this PR. I hope further ideas about what to do will be transferred to some field for developers, after fixing a specific bug. Thank you

felixfontein commented on tests/integration/targets/locale_gen/tasks/non_debian.yml:
```diff
@@ -0,0 +1,52 @@
+          - "glibc-langpack-pl"
+          - "glibc-langpack-en"
+        state: "present"
+        update_cache: "yes"
```
There is usually no reason why the cache should be updated in the middle of a CI run.
```suggestion
```
(Besides that, `"yes"` is a string, not a boolean. `yes` is a boolean, but a form we do not use in this collection. We always use `true` or `false` for YAML booleans.)

felixfontein commented on tests/integration/targets/locale_gen/tasks/non_debian.yml:
```diff
@@ -0,0 +1,52 @@
+      community.general.pacman:
+        name: "glibc-locales"
+        state: "present"
+        update_cache: "yes"
```
```suggestion
```

felixfontein commented on tests/integration/targets/locale_gen/tasks/non_debian.yml:
```diff
@@ -0,0 +1,52 @@
+    owner: "root"
+    group: "root"
+    mode: "0644"
+    force: "yes"
```
```suggestion
    force: true
```

k0ste commented:
Removed preparation tasks for Archlinux. Now somebody need to take a look what with container glibc?

felixfontein commented:
The Arch Linux container had very limited locales (probably to save space), I've fixed that in https://github.com/ansible-community/images/pull/110 and re-triggered the tests; they now pass.

felixfontein commented on tests/integration/targets/locale_gen/tasks/non_debian.yml:
```diff
@@ -0,0 +1,44 @@
+    state: "present"
+  loop: "{{ vars['locale_list_simple'] | flatten(levels=1) }}"
+  loop_control:
+    label: "{{ 'Deploy locale: ' + item['locale'] }}"
```
Please note that this only passes on RHEL 9.5 and Fedora 41 since /etc/locale.gen has been written so that the module thinks the locales are already there. This doesn't really test whether the module works: it will fail if one of the locales is comment out, since it tries to run `locale-gen`, which isn't available on both OSes.

From various places on the net (f.ex. https://www.reddit.com/r/Fedora/comments/v3w393/how_do_i_manually_install_locales/ and https://unix.stackexchange.com/questions/140299/locale-gen-command-in-centos6) it seems that locale-gen was also never available on Fedora/RHEL. So this module likely never worked with these OSes.

Maybe we should kick out the RHEL OS family in that case? Or is there any indication that locale-gen is available on these systems somehow?

k0ste replied:
> it seems that locale-gen was also never available on Fedora/RHEL. So this module likely never worked with these OSes.

Much easier to install package, instead to maintain the locale-gen "thing". The weight of "all languages" in [RHEL is 17Mbytes](https://dl.fedoraproject.org/pub/fedora/linux/updates/41/Everything/x86_64/Packages/g/). Single langpack ~500Kbytes. Don't know why in Debian required for a custom script, still (never used this distro)

> Maybe we should kick out the RHEL OS family in that case? Or is there any indication that locale-gen is available on these systems somehow?

If locale present in `localectl list-locales` then locale ready for use

felixfontein commented:
I'm getting the feeling that this PR is not needed at all (except the test part). RHEL, Fedora, and CentOS (I tried 7 and 8) do not have /etc/locale.gen, do not have the locale-gen CLI tool, so the module never worked on these platforms and this PR isn't changing that.

The reason that this task: https://github.com/k0ste/ansible-role-linux_helper/tree/0f6241d4177a004ed2bbc989d65b89726ab7e0fd/tasks#L10-L16 seemed to have worked before is that the module never did anything, since /etc/locale.gen was templated to the expected value before running the module, which basically made the module a no-op.

Simply removing the community.general.locale_gen module from that task is the right solution to fix the task. That task does not need the module (and never needed it).

felixfontein commented:
> The implementation should [...] use localectl list-locales for any systemd distro in present days IMO

This statement is also not correct. `localectl list-locales` lists the *installed* locales. But this module is actually **doing the installing** on systems that use `locale-gen`. That's the whole purpose of the module. So `localectl list-locales` cannot be used by the module to figure out which locales can be installed.

k0ste commented:
> The reason that this task: https://github.com/k0ste/ansible-role-linux_helper/tree/0f6241d4177a004ed2bbc989d65b89726ab7e0fd/tasks#L10-L16 seemed to have worked before is that the module never did anything, since /etc/locale.gen was templated to the expected value before running the module, which basically made the module a no-op.
> 
> Simply removing the community.general.locale_gen module from that task is the right solution to fix the task. That task does not need the module (and never needed it).

The module does, like said before, `assert_avail` for the locale. I re-implemented the locale tasks (more complex with distro logic) at this week, for possibility work with c9s+ and Ansible 11 https://github.com/k0ste/ansible-role-linux_helper/commit/eadbd25cf99f596d6707cc663c2929df43cdfe8b

russoz commented:
The mechanism it uses to determine _availability_ is to check in the file, in this case `/etc/locale.gen` if there is a line with that locale, **either commented out or not**, the key thing being the comment. If the locale _is commented out_, then it is still available, but not enabled. If it is not commented out (like in the template here), the module understands the locale is _already enabled_, therefore it doesn't need to do anything, therefore it becomes a no-op.

felixfontein commented:
It would probably be helpful to have a separate module `community.general.locale_info` to determine which locales are available (with `locale -a` or `localectl list-locales`). But `community.general.locale_gen` should not be abused for that, it's really only meant for systems supporting `locale-gen` (which does not cover RHEL/CentOS/Fedora and friends).

felixfontein commented:
Since this change doesn't seem necessary I'm closing this. Thanks for your contribution @k0ste!
